### ISSUE TITLE
Loop unrolling support in RyuJIT

### ISSUE THREAD

**Issue by AndreyAkinshin** [dotnet/coreclr#991]
LegacyJIT-x64 can unroll some loops and transform something like
```cs
for (int i = 0; i < 1024; i++)
    Foo(i);
```
to something like
```cs
for (int i = 0; i < 1024; i += 4)
{
    Foo(i);
    Foo(i + 1);
    Foo(i + 2);
    Foo(i + 3);
}
```

Also LegacyJIT-x64 can transform small loops like

```cs
for (int i = 0; i < 4; i++)
    Foo(i);
```
to
```cs
Foo(0);
Foo(1);
Foo(2);
Foo(3);
```

I like this feature because it can increase performance in some cases.

Is it possible to implement loop unrolling in RyuJIT?

See also:
* [RyuJIT CTP5 and loop unrolling](http://aakinshin.net/en/blog/dotnet/ryujit-ctp5-and-loop-unrolling/)
* [Unrolling of small loops in different JIT versions](http://aakinshin.net/en/blog/dotnet/unrolling-of-small-loops-in-different-jit-versions/)

**Comment by mikedn** [comment]
The x86 JIT on which RyuJIT did some unrolling. As far as I can tell the code is still there but it doesn't run - see `optUnrollLoops` in optimizer.cpp. It doesn't run because `optCanCloneLoops` always returns true, probably loop cloning (new to RyuJIT) somehow interferes with the old loop unrolling code. That said, the unrolling done by the x86 JIT isn't great:

```C#
for (int i = 0; i < 3; i++)
    sum += i;
```
generates
```
inc         eax  
inc         eax  
inc         eax  
```

Good loop unrolling isn't trivial and I doubt that the existing unrolling code can be significantly improved.

**Comment by mattwarren** [comment]
> That said, the unrolling done by the x86 JIT isn't great:
```
for (int i = 0; i < 3; i++)
    sum += i;
```
> generates
```
inc         eax  
inc         eax  
inc         eax
```

Forgive the dumb question (I'm trying to learn about the JIT), but what would you expect it to generate? Something like this (or whatever the correct assembly is for adding 3) or is that too much to expect?
```
add eax, 3
```

**Comment by AndreyAkinshin** [comment]
@mikedn, @mattwarren 

Here are the asm listings of the method

```cs
[MethodImpl(MethodImplOptions.NoInlining)]
public int Run()
{
    int sum = 0;
    for (int i = 0; i < 3; i++)
        sum += i;
    return sum;
}
```

for different JIT versions:

**LegacyJIT-x86:**
```nasm
00F33562  in          al,dx  
00F33563  xor         eax,eax  
00F33565  inc         eax  
00F33566  inc         eax  
00F33567  inc         eax  
00F33568  pop         ebp  
00F33569  ret  
```

**LegacyJIT-x64:**
```nasm
00007FF914114470  mov         eax,3  
00007FF914114475  ret  
```

**RyuJIT-x64 RC:**
```nasm
00007FF9140F4230  xor         eax,eax  
00007FF9140F4232  xor         edx,edx  
00007FF9140F4234  add         eax,edx  
00007FF9140F4236  inc         edx  
00007FF9140F4238  cmp         edx,3  
00007FF9140F423B  jl          00007FF9140F4234  
00007FF9140F423D  ret 
```

**Comment by mikedn** [comment]
@mattwarren Yes, `add eax, 3` is expected for the code that I posted. Though the real version includes a `sum = 0` so it's really `mov eax, 3` as in the LegacyJIT-X64 version posted above by @AndreyAkinshin.

And in case that you wonder how come 3 increment instructions were produced: the loop got unrolled as: `sum += 0; sum += 1; sum += 2;`. The first addition was eliminated because it's useless and the last addition was emitted as `inc eax, inc eax` because this sequence is one byte shorter than `add eax, 2`.

@AndreyAkinshin Your LegacyJIT-x86 starts with the wrong instruction `in al,dx`. It's really a `push ebp` but you see `in al,dx` because the VS disassembly window has a bug. Not that it would matter.

**Comment by AndreyAkinshin** [comment]
> It's really a push ebp but you see in al,dx because the VS disassembly window has a bug.

@mikedn, Thanks, it explains a lot!
